File: src/main/java/com/github/svegon/capi/event/input/CrosshairTargetTypeCallback.java
package com.github.svegon.capi.event.input;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.util.hit.HitResult;

import java.util.function.Function;

public interface CrosshairTargetTypeCallback {
    CrosshairTargetTypeCallback EMPTY_LISTENER = HitResult::getType;
    Function<CrosshairTargetTypeCallback[], CrosshairTargetTypeCallback> INVOKER_FACTORY = listeners -> hitResult -> {
        HitResult.Type type = hitResult.getType();

        for (CrosshairTargetTypeCallback listener : listeners) {
            HitResult.Type newType = listener.getCrosshairTargetType(hitResult);

            if (newType != type) {
                return newType;
            }
        }

        return type;
    };

    Event<CrosshairTargetTypeCallback> BLOCK_BREAK = EventFactory.createArrayBacked(CrosshairTargetTypeCallback.class,
            EMPTY_LISTENER, INVOKER_FACTORY);
    Event<CrosshairTargetTypeCallback> ATTACK = EventFactory.createArrayBacked(CrosshairTargetTypeCallback.class,
            EMPTY_LISTENER, INVOKER_FACTORY);
    Event<CrosshairTargetTypeCallback> ITEM_USE = EventFactory.createArrayBacked(CrosshairTargetTypeCallback.class,
            EMPTY_LISTENER, INVOKER_FACTORY);
    Event<CrosshairTargetTypeCallback> ITEM_PICK = EventFactory.createArrayBacked(CrosshairTargetTypeCallback.class,
            EMPTY_LISTENER, INVOKER_FACTORY);

    HitResult.Type getCrosshairTargetType(HitResult hitResult);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/input/CrosshairTargetTypeCallback.kt
package io.github.svegon.capi.event.input

import io.github.svegon.capi.event.input.CrosshairTargetTypeCallback
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.util.hit.HitResult
import java.util.function.Function
import kotlin.Array
import kotlin.invoke

interface CrosshairTargetTypeCallback {
    fun getCrosshairTargetType(hitResult: HitResult?): HitResult.Type?

    companion object {
        val EMPTY_LISTENER: CrosshairTargetTypeCallback = CrosshairTargetTypeCallback { obj: HitResult -> obj.type }
        val INVOKER_FACTORY: Function<Array<CrosshairTargetTypeCallback?>, CrosshairTargetTypeCallback?> =
            Function { listeners: Array<CrosshairTargetTypeCallback?>? ->
                CrosshairTargetTypeCallback { hitResult: HitResult? ->
                    val type: HitResult.Type = hitResult.getType()
                    for (listener in listeners) {
                        val newType: HitResult.Type = listener.getCrosshairTargetType(hitResult)

                        if (newType != type) {
                            return@CrosshairTargetTypeCallback newType
                        }
                    }
                    type
                }
            }

        @JvmField
        val BLOCK_BREAK: Event<CrosshairTargetTypeCallback?> = EventFactory.createArrayBacked(
            CrosshairTargetTypeCallback::class.java,
            EMPTY_LISTENER, INVOKER_FACTORY
        )
        @JvmField
        val ATTACK: Event<CrosshairTargetTypeCallback?> = EventFactory.createArrayBacked(
            CrosshairTargetTypeCallback::class.java,
            EMPTY_LISTENER, INVOKER_FACTORY
        )
        @JvmField
        val ITEM_USE: Event<CrosshairTargetTypeCallback?> = EventFactory.createArrayBacked(
            CrosshairTargetTypeCallback::class.java,
            EMPTY_LISTENER, INVOKER_FACTORY
        )
        @JvmField
        val ITEM_PICK: Event<CrosshairTargetTypeCallback?> = EventFactory.createArrayBacked(
            CrosshairTargetTypeCallback::class.java,
            EMPTY_LISTENER, INVOKER_FACTORY
        )
    }
}
